Scripts/Prerequisite.ps1

```diff
@@ -0,0 +1,96 @@
+param(
+    [Parameter(Mandatory = $true)]
+    [string]$subscriptionId,
+    [Parameter(Mandatory = $true)]
+    [string]$resourceGroupName,
+    [Parameter(Mandatory = $true)]
+    [string]$keyVaultName,
+    [Parameter(Mandatory = $true)]
+    [string]$location,
+    [Parameter(Mandatory = $true)]
+    [string]$certName,
+    [Parameter(Mandatory = $true)]
+    [string]$keyVaultSecretNameForServicePrincipal,
+    [Parameter(Mandatory = $true)]
+    [string]$servicePrincipalName,
+    [Parameter(Mandatory = $true)]
+    [string]$tenantId,
+    [Parameter(Mandatory = $true)]
+    [string]$serviceConnectionName,
+    [Parameter(Mandatory = $true)]
+    [string]$organizationName,
+    [Parameter(Mandatory = $true)]
+    [string]$projectName
+)
+
+try {
+
+  $isLoggedIn = az account show 
+  if ($null -eq $isLoggedIn) {
+      Write-Host "logging In..." -ForegroundColor White
+      az login --use-device-code
+  }
+  
+  Write-Host "Initiating pre-requisites validation..." -ForegroundColor White
+  $subscriptionName = az account list --query "[?id=='$subscriptionId'].name" -o tsv
+  if ($null -eq $subscriptionName) {
+      Write-Warning "You don't have access to subcription: $subscriptionId ."
+      exit
+  }
+  else {
+      Write-Host "Switching to subscription: $subscriptionId" -ForegroundColor White
+      az account set -s $subscriptionId 
+  }
+    
+  $rgAccess = az group show --name $resourceGroupName --query "name" -o tsv
+    
+  if ($null -eq $rgAccess) {
+      Write-Warning "You don't have access to required resource group: $resourceGroupName. please elevate your roles via PIM and try again."
+      exit
+  }
+  Write-Host "Pre-requisites check has completed." -ForegroundColor Green
+  
+  $isKVAvailable  = $(az keyvault list --query "[?name=='$keyVaultName'] | length(@)"  -g $resourceGroupName)
+  if($isKVAvailable -eq 1)
+  {
+    Write-Host "keyvault: $keyVaultName already available ..." -ForegroundColor Green
+  }
+  else
+  {
+    Write-Host "Creating keyvault: $keyVaultName ..." -ForegroundColor Green
+    az keyvault create --name $keyVaultName --resource-group $resourceGroupName --location $location
+  }
+
+  Write-Host "Creating selfsigned certificate in keyvault ..." -ForegroundColor Green
+  az keyvault certificate get-default-policy | Out-File -Encoding utf8 defaultpolicy.json
+  az keyvault certificate create --vault-name $keyVaultName -n $certName  --policy `@defaultpolicy.json
+  az keyvault certificate download --vault-name $keyVaultName -n $certName -f cert.pem
+
+  Write-Host "Creating App registration..." -ForegroundColor Green
+  az ad app create --display-name $servicePrincipalName
+  $appId = az ad app list --display-name $servicePrincipalName --query [0].appId
+  Write-Host "Creating Service Principal..." -ForegroundColor Green
+  az ad sp create --id $appId
```

Comment:
Is appId necessary for creating SP ?

Reply:
yes, for creating sp same appid is  required which used for App registration